### PR TITLE
chore: auto doc reference endpoints

### PR DESCRIPTION
**Pull request by ryanfang5**

### CODE REVIEW

**Comment by ryanfang5** [comment]
I am not sure that the example responses for "updates" are correct.

File: tests/references/test_api.py [hunk 1]
```diff
@@ -280,22 +280,15 @@ async def test_edit(data_type, error, mocker, snapshot, fake2, spawn_client, res
         )
         return
 
-    if error == "422":
-        await resp_is.invalid_input(
```
Comment (igboyes) [inline tests/references/test_api.py:284]: Missing check of status code. I suggest having two error tests like `400_duplicates`, `400_invalid_input`.

File: virtool/references/api.py [hunk 2]
```diff
@@ -52,608 +68,723 @@
...
+        )
 
-    if release is None:
-        raise HTTPBadRequest(text="Target release does not exist")
 
-    created_at = virtool.utils.timestamp()
+@routes.view("/refs/{ref_id}")
+@routes.jobs_api.get("/refs/{ref_id}")
+class ReferenceView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceResponse], r404]:
```
Comment (igboyes) [inline virtool/references/api.py:244]: This should return `403` as well if the client doesn't have read permission on the reference.

File: virtool/references/api.py [hunk 3]
```diff
@@ -52,608 +68,723 @@
...
 
-    ref_id = req.match_info["ref_id"]
 
-    if not await virtool.mongo.utils.id_exists(db.references, ref_id):
-        raise NotFound()
+@routes.view("/refs/{ref_id}/otus")
+class ReferenceOtusView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceOTUResponse], r404]:
+        """
+        Get OTUs of a reference
```
Comment (igboyes) [inline virtool/references/api.py:451]:
```suggestion
        Find OTUs.
        
        Finds OTUs by name or abbreviation. Results are paginated.
```

File: virtool/references/api.py [hunk 4]
```diff
@@ -52,608 +68,723 @@
...
-    headers = {"Content-Location": f"/tasks/{task['id']}"}
+        return json_response(data)
 
-    return json_response(task, 202, headers)
 
+@routes.view("/refs/{ref_id}/indexes")
+class ReferenceIndexesView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceIndexResponse], r404]:
+        """
+        List indexes of the reference.
```
Comment (igboyes) [inline virtool/references/api.py:511]:
```suggestion
        List indexes.
        
        Retrieves a paginated list of indexes that have been created for the reference.
```

File: virtool/references/api.py [hunk 5]
```diff
@@ -52,608 +68,723 @@
...
-        {"_id": ref_id, "groups.id": group_id}, ["groups", "users"]
-    )
 
-    if document is None:
-        raise NotFound()
+@routes.view("/refs/{ref_id}/groups")
+class ReferenceGroupsView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceGroupsResponse], r404]:
+        """
+        List groups of the reference.
```
Comment (igboyes) [inline virtool/references/api.py:534]:
```suggestion
        List groups.
        
        Lists all groups that have access to the reference.
```

File: virtool/references/api.py [hunk 6]
```diff
@@ -52,608 +68,723 @@
...
-    document = await db.references.find_one(ref_id, ["groups", "users"])
+        return json_response(groups)
 
-    if document is None:
-        raise NotFound()
+    async def post(
+        self, data: CreateReferenceGroupsSchema
+    ) -> Union[r201[CreateReferenceGroupResponse], r400, r403, r404]:
+        """
+        Allow groups to view, use and modify the reference.
```
Comment (igboyes) [inline virtool/references/api.py:556]:
```suggestion
        Add a group.
        
        Adds a group to the reference. Groups can view, use, and modify the reference.
```

File: virtool/references/api.py [hunk 7]
```diff
@@ -52,608 +68,723 @@
...
-@schema({**RIGHTS_SCHEMA, "user_id": {"type": "string", "required": True}})
-async def add_user(req):
-    db = req.app["db"]
-    data = req["data"]
-    ref_id = req.match_info["ref_id"]
+@routes.view("/refs/{ref_id}/groups/{group_id}")
+class ReferenceGroupView(PydanticView):
+    async def get(self) -> Union[r200[ReferenceGroupResponse], r404]:
+        """
+        Get a specific group of the reference
```
Comment (igboyes) [inline virtool/references/api.py:600]:
```suggestion
        Get a group.
        
        Retrieves the details of a group that has access to the reference.
```

File: virtool/references/api.py [hunk 8]
```diff
@@ -52,608 +68,723 @@
...
-    subdocument = await virtool.references.db.edit_group_or_user(
-        db, ref_id, group_id, "groups", data
-    )
+@routes.view("/refs/{ref_id}/users")
+class ReferenceUsersView(PydanticView):
+    async def post(
+        self, data: CreateReferenceUsersSchema
+    ) -> Union[r200[List[ReferenceUsersSchema]], r400, r403, r404]:
+        """
+        Allow users to view, use and modify the reference.
```
Comment (igboyes) [inline virtool/references/api.py:688]: Update to match docstring for groups.

File: virtool/references/api.py [hunk 9]
```diff
@@ -52,608 +68,723 @@
...
-    db = req.app["db"]
-    ref_id = req.match_info["ref_id"]
-    group_id = req.match_info["group_id"]
+@routes.view("/refs/{ref_id}/users/{user_id}")
+class ReferenceUserView(PydanticView):
+    async def patch(
+        self, data: ReferenceRightsSchema
+    ) -> Union[r200[ReferenceGroupResponse], r403, r404]:
+        """
+        Edit the reference rights of a specific user.
```
Comment (igboyes) [inline virtool/references/api.py:734]: Update to match docstring for groups.

File: virtool/references/api.py [hunk 10]
```diff
@@ -52,608 +68,723 @@
...
-    await virtool.references.db.delete_group_or_user(db, ref_id, group_id, "groups")
+        if subdocument is None:
+            raise NotFound()
 
-    raise HTTPNoContent
+        return json_response(await extend_user(db, subdocument))
 
+    async def delete(self) -> Union[r204, r403, r404]:
+        """
+        Remove the user from the reference.
```
Comment (igboyes) [inline virtool/references/api.py:767]: Update to match docstring for groups.

File: virtool/references/api.py [hunk 11]
```diff
@@ -52,608 +68,723 @@
...
+        if document is not None:
+            for group in document.get("groups", []):
+                if group["id"] == group_id:
+                    return json_response(group)
+
+    async def patch(
+        self, data: ReferenceRightsSchema
+    ) -> Union[r200[ReferenceGroupResponse], r403, r404]:
+        """
+        Edit the reference rights of a specific group.
```
Comment (igboyes) [inline virtool/references/api.py:626]:
```suggestion
        Update a group.
        
        Updates the access rights a group has on the reference.
```

File: virtool/references/api.py [hunk 12]
```diff
@@ -52,608 +68,723 @@
...
 
-@routes.delete("/refs/{ref_id}")
-async def remove(req):
-    """
-    Delete a reference and its otus, history, and indexes.
+@routes.view("/refs/{ref_id}/history")
+class ReferenceHistoryView(PydanticView):
+    async def get(self) -> Union[r200[GetHistoryResponse], r404]:
+        """
+        List history of the reference.
```
Comment (igboyes) [inline virtool/references/api.py:479]:
```suggestion
        List history.
        
        Retrieves a paginated list of changes made to OTUs in the reference.
```

File: virtool/references/api.py [hunk 13]
```diff
@@ -52,608 +68,723 @@
...
+        if updates is not None:
+            updates.reverse()
 
-        task = await req.app["tasks"].add(ImportReferenceTask, context=context)
+        return json_response(updates or [])
 
-        document["task"] = {"id": task["id"]}
+    async def post(self) -> Union[r201[CreateReferenceUpdateResponse], r403, r404]:
+        """
+        Update the reference using the linked remote reference.
```
Comment (igboyes) [inline virtool/references/api.py:398]:
```suggestion
        Update a reference.
        
        Updates the reference to the last version of the linked remote reference.
```

File: virtool/references/api.py [hunk 14]
```diff
@@ -52,608 +68,723 @@
...
-        document["task"] = {"id": task["id"]}
 
-    elif import_from:
-        if not await get_row(pg, Upload, ("name_on_disk", import_from)):
-            raise NotFound("File not found")
+@routes.view("/refs/{ref_id}/updates")
+class ReferenceUpdatesView(PydanticView):
+    async def get(self) -> r200[GetReferenceUpdateResponse]:
+        """
+        List all updates made to the reference.
```
Comment (igboyes) [inline virtool/references/api.py:376]:
```suggestion
        List updates.
        
        Lists all updates made to the reference.
```

File: virtool/references/api.py [hunk 15]
```diff
@@ -52,608 +68,723 @@
...
-    release_id = data.get("release_id") or 11447367
 
-    if clone_from:
-        if not await db.references.count_documents({"_id": clone_from}):
-            raise HTTPBadRequest(text="Source reference does not exist")
+@routes.view("/refs/{ref_id}/release")
+class ReferenceReleaseView(PydanticView):
+    async def get(self) -> r200[ReferenceReleaseResponse]:
+        """
+        Get the latest update from GitHub and return it.
```
Comment (igboyes) [inline virtool/references/api.py:339]:
```suggestion
        Get latest update.
        
        Retrieves the latest remote reference update from GitHub.
```

File: virtool/references/oas.py [hunk 16]
```diff
@@ -0,0 +1,917 @@
+from typing import Optional, Union, List
```
Comment (igboyes) [inline virtool/references/oas.py:1]: Excellent use of validators in this module. But... we do need to test them. Add a `test_oas.py` to make sure that model implementing the validators work as expected.

**Review by igboyes — CHANGES_REQUESTED** [review]

**Comment by ryanfang5** [comment]
I also added Permission.create_ref to the post method for /refs. However, I noticed that this method also requires Permission.upload_file but only when importing from a file. I am not sure how to implement this since the permissions are checked before the method is run.

File: tests/references/test_oas.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+@pytest.mark.parametrize(
+    "error", [None, "data_type", "remote_from", "two_values", "all_values"]
+)
+def test_schema(error, mock_create_reference_schema):
+    CreateReferenceSchema(**mock_create_reference_schema)
+
+    if error == "data_type":
+        with pytest.raises(ValidationError):
+            mock_create_reference_schema.update({"data_type": "bad type"})
```
Comment (igboyes) [inline tests/references/test_oas.py:25]: Put the update dictionary as the pytest parameter instead. This test can be heavily simplified.

**Review by igboyes — CHANGES_REQUESTED** [review]

**Comment by igboyes** [comment]
> I also added Permission.create_ref to the post method for /refs. However, I noticed that this method also requires Permission.upload_file but only when importing from a file. I am not sure how to implement this since the permissions are checked before the method is run.

Are you sure? I don't see that in the existing code. I believe the upload is executed in a separate request before the creation request is made.

**Comment by ryanfang5** [comment]
> > I also added Permission.create_ref to the post method for /refs. However, I noticed that this method also requires Permission.upload_file but only when importing from a file. I am not sure how to implement this since the permissions are checked before the method is run.
> 
> Are you sure? I don't see that in the existing code. I believe the upload is executed in a separate request before the creation request is made.

My mistake, you are right.

**Comment by igboyes** [comment]
> My mistake, you are right.

Perfect! Can you resolve conflict and I will merge?

**Review by igboyes — APPROVED** [review]

**Pull request merged**